### PR TITLE
Coverage changes

### PR DESCRIPTION
**Pull request by thk123**

### CODE REVIEW

**Comment by peterschrammel** [comment]
@thk123, linting and squashing please.

There's a compilation error. For some reason also cegis links with goto-instrument/cover.o (not sure why this is actually needed).

File: src/goto-instrument/cover.cpp [hunk 1]
```diff
@@ -98,6 +100,174 @@ class basic_blockst
...
+    file=(*it)["file"].value;
+    source_location.set_file(file);
+
+    //get the function of each existing goal
+    function=(*it)["function"].value;
+    source_location.set_function(function);
+
+    //get the lines array
+    if ((*it)["lines"].is_array())
+	{
```
Comment (peterschrammel) [inline src/goto-instrument/cover.cpp]: weird indent

File: src/goto-instrument/cover.cpp [hunk 2]
```diff
@@ -98,6 +100,174 @@ class basic_blockst
...
+  if(!json.is_array())
+  {
+    messaget message(message_handler);
+    message.error() << "expecting an array in the " <<  coverage << " file, but got "
+                    << json << messaget::eom;
+    exit(0);
+  }
+
+  irep_idt file, function, line;
+  for(jsont::arrayt::const_iterator
```
Comment (peterschrammel) [inline src/goto-instrument/cover.cpp]: for each

File: src/goto-instrument/cover.cpp [hunk 3]
```diff
@@ -98,6 +100,174 @@ class basic_blockst
...
+  source_locationt source_location;
+  goals.set_no_trivial_tests(false);
+
+  //check coverage file
+  if(parse_json(coverage, message_handler, json))
+  {
+    messaget message(message_handler);
+    message.error() << coverage << " file is not a valid json file"
+                    << messaget::eom;
+    exit(0);
```
Comment (peterschrammel) [inline src/goto-instrument/cover.cpp]: exiting with 0 is not a good idea when the reason is an error. Use exit(6). (Several instances)

File: src/goto-instrument/cover.cpp [hunk 4]
```diff
@@ -98,6 +100,174 @@ class basic_blockst
...
+    source_location.set_file(file);
+
+    //get the function of each existing goal
+    function=(*it)["function"].value;
+    source_location.set_function(function);
+
+    //get the lines array
+    if ((*it)["lines"].is_array())
+	{
+	  for(jsont::arrayt::const_iterator
```
Comment (peterschrammel) [inline src/goto-instrument/cover.cpp]: for each

File: src/goto-instrument/cover.cpp [hunk 5]
```diff
@@ -98,6 +100,174 @@ class basic_blockst
...
+	      itg=(*it)["lines"].array.begin();
+	      itg!=(*it)["lines"].array.end();
+	      itg++)
+	  {
+        //get the line of each existing goal
+        line=(*itg)["number"].value;
+        source_location.set_line(line);
+        goals.set_goals(source_location);
+	  }
+	}
```
Comment (peterschrammel) [inline src/goto-instrument/cover.cpp]: weird indent (many instances... check for tabs)

File: src/goto-instrument/cover.cpp [hunk 6]
```diff
@@ -98,6 +100,174 @@ class basic_blockst
...
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+bool coverage_goalst::is_existing_goal(source_locationt source_location)
+{
+  std::vector<source_locationt>::iterator it = existing_goals.begin();
```
Comment (peterschrammel) [inline src/goto-instrument/cover.cpp]: cpplint

File: src/goto-instrument/cover.cpp [hunk 7]
```diff
@@ -98,6 +100,174 @@ class basic_blockst
...
+  while (it!=existing_goals.end())
+  {
+    if (!source_location.get_file().compare(it->get_file()) &&
+	!source_location.get_function().compare(it->get_function()) &&
+	!source_location.get_line().compare(it->get_line()))
+      break;
+    ++it;
+  }
+  if(it == existing_goals.end())
```
Comment (peterschrammel) [inline src/goto-instrument/cover.cpp]: cpplint

File: src/goto-instrument/cover.cpp [hunk 8]
```diff
@@ -1437,6 +1611,81 @@ void instrument_cover_goals(
...
+
+void instrument_cover_goals(
+  const symbol_tablet &symbol_table,
+  goto_functionst &goto_functions,
+  coverage_criteriont criterion)
+{
+  Forall_goto_functions(f_it, goto_functions)
+  {
+    if(f_it->first==ID__start ||
+       f_it->first=="__CPROVER_initialize")
```
Comment (peterschrammel) [inline src/goto-instrument/cover.cpp]: use util/cprover_prefix.h

File: src/goto-instrument/cover.cpp [hunk 9]
```diff
@@ -1437,6 +1611,81 @@ void instrument_cover_goals(
...
+  //check whether this is a constructor/destructor or a get/set (pattern)
+  if (!count_goto && !count_assignments && !count_decl)
+    result=false;
+  else
+    result = !((count_decl==0) && (count_goto<=1) &&
+             (count_assignments>0 && count_assignments<5));
+
+  return result;
 }
+
```
Comment (peterschrammel) [inline src/goto-instrument/cover.cpp]: remove blank line

**Comment by peterschrammel** [comment]
Do some squashing of these commits as most of them seem to be just fixes of previous commits.

**Comment by thk123** [comment]
@lucasccordeiro Are you OK to do these changes?

**Comment by lucasccordeiro** [comment]
@thk123: I can implement @peterschrammel, but you need to grant me access to commit in your branch. I started implementing @peterschrammel's comments, but I'm getting the following error message:

````
Lucass-MacBook-Pro:src lucasccordeiro$ git push origin feature/filtered-lucas-changes
remote: Permission to thk123/cbmc.git denied to lucasccordeiro.
fatal: unable to access 'https://github.com/thk123/cbmc.git/': The requested URL returned error: 403
````

File: src/goto-instrument/cover.cpp (hunk 1 above)
Comment (lucasccordeiro) [inline src/goto-instrument/cover.cpp]: Implemented as suggested.

File: src/goto-instrument/cover.cpp (hunk 4 above)
Comment (lucasccordeiro) [inline src/goto-instrument/cover.cpp]: Implemented as suggested.

File: src/goto-instrument/cover.cpp (hunk 5 above)
Comment (lucasccordeiro) [inline src/goto-instrument/cover.cpp]: I have applied cpplint to that code in cover.cpp

File: src/goto-instrument/cover.cpp (hunk 6 above)
Comment (lucasccordeiro) [inline src/goto-instrument/cover.cpp]: I have applied cpplint as suggested.

File: src/goto-instrument/cover.cpp (hunk 7 above)
Comment (lucasccordeiro) [inline src/goto-instrument/cover.cpp]: I have applied cpplint as suggested.

File: src/goto-instrument/cover.cpp (hunk 8 above)
Comment (lucasccordeiro) [inline src/goto-instrument/cover.cpp]: Implemented as suggested.

File: src/goto-instrument/cover.cpp (hunk 9 above)
Comment (lucasccordeiro) [inline src/goto-instrument/cover.cpp]: Implemented as suggested.

**Comment by lucasccordeiro** [comment]
@peterschrammel: I have implemented all your comments as suggested in this PR.

**Comment by peterschrammel** [comment]
@thk123, @lucasccordeiro: The dependency of cegis on json.a has not been fixed yet: https://travis-ci.org/diffblue/cbmc/jobs/211012246#L2365
Please fix.

**Pull request merged**